osquery/tables/system/windows/prefetch.cpp

```diff
@@ -0,0 +1,408 @@
+
+namespace osquery {
+namespace tables {
+namespace {
+
+const std::string kPrefetchLocation = (getSystemRoot() / "Prefetch\\").string();
+const unsigned int kPrefetchVersionWindows10 = 30;
+const unsigned int kPrefetchVersionWindows8 = 26;
+const unsigned int kPrefetchVersionWindows7 = 23;
+const unsigned int kPrefetchVolumeVersionWindows10 = kPrefetchVersionWindows10;
+const unsigned int kPrefetchVolumeSizeWindows10 = 96;
+const unsigned int kPrefetchVolumeSizeWindows8 = 104;
+
+struct PrefetchHeader {
+  std::uint32_t file_size;
+  std::string filename;
+  std::string prefetch_hash;
+};
+
+struct PrefetchFileInfo {
+  std::string files;
+  size_t files_count;
+  LONGLONG last_run_time;
+  std::string run_times;
+  size_t run_count;
+};
+
+struct PrefetchVolumeInfo {
+  std::string directories;
+  size_t directories_count;
+  std::string volume_serial;
+  std::string volume_creation;
+};
+
+#pragma pack(push, 1)
+typedef struct _PREFETCH_COMPRESSED_HEADER {
+  DWORD Signature;
+  DWORD TotalUncompressedSize;
+  BYTE CompressedData[1]; // arbitrary size
+} PREFETCH_COMPRESSED_HEADER, *PPREFETCH_COMPRESSED_HEADER;
+
+typedef struct _PREFETCH_FILE_HEADER {
+  DWORD Version;
+  DWORD Signature;
+  DWORD Reserved1;
+  DWORD FileSize;
+  WCHAR FileName[30];
+  DWORD Hash;
+  DWORD Reserved2;
+} PREFETCH_FILE_HEADER, *PPREFETCH_FILE_HEADER;
+
+typedef struct _PREFETCH_FILE_INFORMATION {
+  DWORD FileMetricsArrayOffset;
+  DWORD NumberOfMetricEntries;
+  DWORD TraceChainsArrayOffset;
+  DWORD NumberOfTraceChains;
+  DWORD FileNameStringsOffset;
+  DWORD FileNameStringsSize;
+  DWORD VolumeInformationOffset;
+  DWORD NumberOfVolumes;
+  DWORD VolumesInformationSize;
+  union {
+    struct {
+      DWORD Reserved1[2];
+      FILETIME LastRunTime;
+      DWORD Reserved2[4];
+      DWORD RunCount;
+    } v23;
+    struct {
+      DWORD Reserved1[2];
+      FILETIME LastRunTime;
+      FILETIME OtherRunTimes[7];
+      DWORD Reserved2[4];
+      DWORD RunCount;
+    } v26;
+    struct {
+      DWORD Reserved1[2];
+      FILETIME LastRunTime;
+      FILETIME OtherRunTimes[7];
+      DWORD Reserved2[2];
+      DWORD RunCount;
+    } v30v2;
+  } ext;
+} PREFETCH_FILE_INFORMATION, *PPREFETCH_FILE_INFORMATION;
+
+typedef struct _PREFETCH_VOLUME_INFORMATION {
+  DWORD VolumePathOffset;
+  DWORD VolumeDevicePathNumberOfCharacters;
+  FILETIME VolumeCreationTime;
+  DWORD VolumeSerialNumber;
+  DWORD FileReferencesOffset;
+  DWORD FileReferencesDataSize;
+  DWORD DirectoryStringsOffset;
+  DWORD NumberOfDirectoryStrings;
+  DWORD Reserved1;
+} PREFETCH_VOLUME_INFORMATION, *PPREFETCH_VOLUME_INFORMATION;
+
+typedef struct _DIRECTORY_STRING {
+  USHORT Size;
+  WCHAR Directory[1]; // arbitrary size
+} DIRECTORY_STRING, *PDIRECTORY_STRING;
+#pragma pack(pop)
+
+} // namespace
+
+PrefetchHeader parseHeader(const PREFETCH_FILE_HEADER* header) {
+  PrefetchHeader result;
+  if (header->FileName[(sizeof(header->FileName) / sizeof(WCHAR)) - 1] ==
+      '\0') {
```

Comment:
Does this also have to be `L'\0'`?

Reply:
Yes, preferred. Also, in this specific case, substituting with the macro `ARRAYSIZE(header->FileName) - 1` instead of `(sizeof(header->FileName) / sizeof(WCHAR)) - 1` might be more clear.

Hmmm, in this case, if `FileName` is not NULL-terminated, `result.filename` remains empty. We ought to log these cases if we encounter them since this is not expected.

Reply:
added L, switched to ARRAYSIZE.
I wasnt really sure if logging this is good? It feels a little too informational/debugging? Im not 100% what osquery considers appropriate logging vs excessive logging.
Although u have a good point that it if the filename is not NULL terminated then that would be unexpected and could be worth logging.
Added LOG(INFO)